File: Data Structures & Algorithms/balanced-binary-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        def height(node: Optional[TreeNode]):
            if node:
                l = height(node.left)
                r = height(node.right)
                return 1+max(l,r)
            return 0
        if not root :
            return True
        if root:
            diff = height(root.left)-height(root.right)
        else :
            diff = 0 
        if abs(diff) > 1:
            return False
        return self.isBalanced(root.left) and self.isBalanced(root.right)

File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool trav(TreeNode * root, TreeNode * subRoot){
        if(!root and !subRoot) return true;
        if(!root or !subRoot) return false;
        if(root->val == subRoot->val){
            return trav(root->left, subRoot->left) and trav(root->right, subRoot->right);
        }
        return false;
    }
    bool isSubtree(TreeNode* root, TreeNode* subRoot) {
        //traverse the tree till the first node matches then start
        if(!root || !subRoot) return false;
        if(trav(root,subRoot)){
            return true;
        }
        return isSubtree(root->left, subRoot) || isSubtree(root->right, subRoot);

    }
};
